### ISSUE TITLE
Failure to find class defined in an other compilation unit

### ISSUE THREAD

**Issue by smarter** [lampepfl/dotty#850]
Put the following in `A.scala`:
```scala
import B.hello

class A(private[this] var foo: Int)
```
And the following in `B.scala`:
```scala
object B {
  def hello: Int = 1
}
```
Then this fails:
```scala
% dotc A.scala B.scala
A.scala:1: error: not found: B
import B.hello
       ^
one error found
```
This does not happen if we replace the `var` by a `val`, or `private[this]` by `private` or nothing

**Comment by DarkDimius** [comment]
>note that dotc in master is currently broken, you need to have dotty-staging@1adc563 applied

Not entirely broken, it is just unconditionally bootstrapping itself. If you let it finish once, it should be fine.

**Comment by smarter** [comment]
OK, I've tried to investigate this and here's what seems to happen:

1. `dotc A.scala B.scala`
2. Enter the compilation unit corresponding to `A.scala`
3. Call `ClassCompleter#init()` on `A`
4. Expand `private[this] var foo: Int` because it's a parameter accessor
5.  In `Desugar#valDef`, check `ctx.owner is Trait`, this only happens because the parameter accessor is `private[this] var`:
 ```scala
(mods is Mutable) && ctx.owner.isClass && (!(mods is PrivateLocal) || (ctx.owner is Trait))
```

6.  Force completion of `class A`
7. Type the constructor of `A`
8. Resolve the name `Int` in `typedIdent#findRef`
9. Complete the type of `import B.hello`
10. Type `B`
11. Fail to type `B` because it's not in scope.

Notice that steps 5 to 11 happen only when a class has a `private[this] var` accessor, otherwise the class should not be completed during the enter phase.

@odersky : What do you think? Is it normal to complete classes when entering symbols or should it be avoided?

**Comment by DarkDimius** [comment]
> `ctx.owner is Trait`

Could you try `ctx.owner.flagsUNSAFE is Trait`? AFAIK trait flag is set even before type is completed

**Comment by smarter** [comment]
That works yes, thanks! `Trait` seems to be set in the parser so this should work OK, but I still would like to know if completing the class is a bug or not. If it's a bug we should add some checks to never complete classes when entering symbols.

**Comment by DarkDimius** [comment]
I believe it competing symbol in namer is a bug. AFAIK completing a symbol relies on assumption that all top level definitions are already entered.

**Comment by odersky** [comment]
Generally, we should always complete as little as possible. But we should
not use flagsUNSAFE. Instea, if trait is already set in parser we should
include Trait in Flags.FromStartFlags, which will mean we can test for it
without forcing.


On Fri, Oct 23, 2015 at 10:28 PM, Dmitry Petrashko <notifications@github.com
> wrote:

> I believe it competing symbol in namer is a bug. AFAIK completing a symbol
> relies on assumption that all top level definitions are already entered.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/lampepfl/dotty/issues/850#issuecomment-150681480>.
>



-- 
Martin Odersky
EPFL

**Comment by smarter** [comment]
Interesting, I wasn't aware of `FromStartFlags`, but adding `Trait` to it fails because of this assert:
```scala
assert(FromStartFlags.isTermFlags && FromStartFlags.isTypeFlags)
```
Why is this check needed?

**Comment by odersky** [comment]
I guess this was a precaution because otherwise we might try to have a term
flag & a type flag in the set and the result would be an illegal flag set
(which would give a runtime exception but a harder to diagnose one). But
you can try to just drop the assert.

 - Martin


On Fri, Oct 23, 2015 at 10:36 PM, Guillaume Martres <
notifications@github.com> wrote:

> Interesting, I wasn't aware of FromStartFlags, but adding Trait to it
> fails because of this assert:
>
> assert(FromStartFlags.isTermFlags && FromStartFlags.isTypeFlags)
>
> Why is this check needed?
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/lampepfl/dotty/issues/850#issuecomment-150683211>.
>



-- 
Martin Odersky
EPFL

**Comment by DarkDimius** [comment]
I am hesitating to drop this assert.

This assert is there for a reason: `<=` is not aware of difference between TermFlags and TypeFlags and works simply in terms of subsets: ` (bits & that.bits) == bits`. If eiterh

**Comment by smarter** [comment]
Dropping the assert and adding `Trait` makes everything explode, no clue why:
```scala
[info] Running dotty.tools.dotc.Main -Ylog:all try/simple.scala
[info] exception caught when loading module class org: dotty.tools.dotc.core.Types$CyclicReference: cyclic reference involving module org
[info] error while completing module class org#4
[info] error while completing module org#3
[info] exception caught when loading module class <root>: dotty.tools.dotc.core.Types$CyclicReference: cyclic reference involving module org
[info] error while completing module class <root>#2
[info] error while completing module <root>#1
[error] Exception in thread "main" dotty.tools.dotc.core.Types$CyclicReference: cyclic reference involving module org
[error]         at dotty.tools.dotc.core.Types$CyclicReference$.apply(Types.scala:3235)
[error]         at dotty.tools.dotc.core.SymDenotations$SymDenotation.completeFrom(SymDenotations.scala:166)
[error]         at dotty.tools.dotc.core.SymDenotations$SymDenotation.info(SymDenotations.scala:152)
[error]         at dotty.tools.dotc.core.SymDenotations$SymDenotation.ensureCompleted(SymDenotations.scala:267)
[error]         at dotty.tools.dotc.core.SymDenotations$SymDenotation.flags(SymDenotations.scala:100)
[error]         at dotty.tools.dotc.core.SymDenotations$SymDenotation.is(SymDenotations.scala:128)
[error]         at dotty.tools.dotc.core.Symbols$Symbol.isFresh(Symbols.scala:409)
[error]         at dotty.tools.dotc.core.Types$TermRef$.withSymAndName(Types.scala:1709)
[error]         at dotty.tools.dotc.core.Types$TermRef$.apply(Types.scala:1680)
[error]         at dotty.tools.dotc.core.Types$Type.select(Types.scala:914)
[error]         at dotty.tools.dotc.core.SymbolLoaders$PackageLoader.doComplete(SymbolLoaders.scala:157)
[error]         at dotty.tools.dotc.core.SymbolLoader.complete(SymbolLoaders.scala:205)
[error]         at dotty.tools.dotc.core.SymDenotations$SymDenotation.completeFrom(SymDenotations.scala:170)
[error]         at dotty.tools.dotc.core.SymDenotations$SymDenotation.info(SymDenotations.scala:152)
[error]         at dotty.tools.dotc.core.SymDenotations$SymDenotation.ensureCompleted(SymDenotations.scala:267)
[error]         at dotty.tools.dotc.core.SymDenotations$SymDenotation.flags(SymDenotations.scala:100)
[error]         at dotty.tools.dotc.core.SymDenotations$ModuleCompleter.complete(SymDenotations.scala:1792)
[error]         at dotty.tools.dotc.core.SymDenotations$SymDenotation.completeFrom(SymDenotations.scala:170)
[error]         at dotty.tools.dotc.core.SymDenotations$SymDenotation.info(SymDenotations.scala:152)
[error]         at dotty.tools.dotc.core.SymDenotations$SymDenotation.ensureCompleted(SymDenotations.scala:267)
[error]         at dotty.tools.dotc.core.SymDenotations$SymDenotation.flags(SymDenotations.scala:100)
[error]         at dotty.tools.dotc.core.SymDenotations$SymDenotation.is(SymDenotations.scala:128)
[error]         at dotty.tools.dotc.core.Symbols$Symbol.entered(Symbols.scala:425)
[error]         at dotty.tools.dotc.core.SymbolLoaders.enterPackage(SymbolLoaders.scala:86)
[error]         at dotty.tools.dotc.core.SymbolLoaders$PackageLoader$$anonfun$doComplete$4.apply(SymbolLoaders.scala:170)
[error]         at dotty.tools.dotc.core.SymbolLoaders$PackageLoader$$anonfun$doComplete$4.apply(SymbolLoaders.scala:169)
[error]         at scala.collection.Iterator$class.foreach(Iterator.scala:750)
[error]         at scala.collection.AbstractIterator.foreach(Iterator.scala:1202)
[error]         at scala.collection.IterableLike$class.foreach(IterableLike.scala:72)
[error]         at scala.collection.AbstractIterable.foreach(Iterable.scala:54)
[error]         at dotty.tools.dotc.core.SymbolLoaders$PackageLoader.doComplete(SymbolLoaders.scala:169)
[error]         at dotty.tools.dotc.core.SymbolLoader.complete(SymbolLoaders.scala:205)
[error]         at dotty.tools.dotc.core.SymDenotations$SymDenotation.completeFrom(SymDenotations.scala:170)
[error]         at dotty.tools.dotc.core.SymDenotations$SymDenotation.info(SymDenotations.scala:152)
[error]         at dotty.tools.dotc.core.SymDenotations$SymDenotation.ensureCompleted(SymDenotations.scala:267)
[error]         at dotty.tools.dotc.core.SymDenotations$SymDenotation.flags(SymDenotations.scala:100)
[error]         at dotty.tools.dotc.core.SymDenotations$ModuleCompleter.complete(SymDenotations.scala:1792)
[error]         at dotty.tools.dotc.core.SymDenotations$SymDenotation.completeFrom(SymDenotations.scala:170)
[error]         at dotty.tools.dotc.core.SymDenotations$SymDenotation.info(SymDenotations.scala:152)
[error]         at dotty.tools.dotc.core.SymDenotations$SymDenotation.ensureCompleted(SymDenotations.scala:267)
[error]         at dotty.tools.dotc.core.SymDenotations$SymDenotation.flags(SymDenotations.scala:100)
[error]         at dotty.tools.dotc.core.SymDenotations$SymDenotation.is(SymDenotations.scala:128)
[error]         at dotty.tools.dotc.core.SymDenotations$SymDenotation.moduleClass(SymDenotations.scala:722)
[error]         at dotty.tools.dotc.core.Definitions.RootClass$lzycompute(Definitions.scala:98)
[error]         at dotty.tools.dotc.core.Definitions.RootClass(Definitions.scala:97)
[error]         at dotty.tools.dotc.core.Denotations$DenotationsBase$class.recur$1(Denotations.scala:987)
[error]         at dotty.tools.dotc.core.Denotations$DenotationsBase$class.staticRef(Denotations.scala:1003)
[error]         at dotty.tools.dotc.core.Contexts$ContextBase.staticRef(Contexts.scala:485)
[error]         at dotty.tools.dotc.core.Symbols$class.requiredPackage(Symbols.scala:340)
[error]         at dotty.tools.dotc.core.Contexts$Context.requiredPackage(Contexts.scala:53)
[error]         at dotty.tools.dotc.core.Definitions.ScalaPackageVal$lzycompute(Definitions.scala:110)
[error]         at dotty.tools.dotc.core.Definitions.ScalaPackageVal(Definitions.scala:110)
[error]         at dotty.tools.dotc.core.Definitions.ScalaPackageClass$lzycompute(Definitions.scala:112)
[error]         at dotty.tools.dotc.core.Definitions.ScalaPackageClass(Definitions.scala:112)
[error]         at dotty.tools.dotc.core.Definitions.AnyClass$lzycompute(Definitions.scala:141)
[error]         at dotty.tools.dotc.core.Definitions.AnyClass(Definitions.scala:141)
[error]         at dotty.tools.dotc.core.Definitions.syntheticCoreClasses$lzycompute(Definitions.scala:630)
[error]         at dotty.tools.dotc.core.Definitions.syntheticCoreClasses(Definitions.scala:629)
[error]         at dotty.tools.dotc.core.Definitions.init(Definitions.scala:652)
[error]         at dotty.tools.dotc.Compiler.rootContext(Compiler.scala:102)
[error]         at dotty.tools.dotc.Compiler.newRun(Compiler.scala:129)
[error]         at dotty.tools.dotc.Driver.doCompile(Driver.scala:20)
[error]         at dotty.tools.dotc.Driver.process(Driver.scala:44)
[error]         at dotty.tools.dotc.Driver.main(Driver.scala:48)
[error]         at dotty.tools.dotc.Main.main(Main.scala)
```

**Comment by DarkDimius** [comment]
@smarter see explanation on why it explodes in my comment.

**Comment by DarkDimius** [comment]
@odersky 

>otherwise we might try to have a term flag & a type flag in the set and the result would be an illegal flag set

FromStartFlags needs to be have both TERM and TYPE flag set and needs to be illegal from that point of view. Otherwise `fs <= FromStartFlags` is going to be either always `false` for Terms or always `false` for Types.

**Comment by smarter** [comment]
I see, so can we check if something is a type or a term without completing it? If so we could rename `FromStartFlags` to `FromStartCommonFlags` and add `FromStartTypeFlags`.

**Comment by DarkDimius** [comment]
@smarter could you check if `lazy` flag is set by parser? 
If it is, you can define 
```scala
final val LazyOrTrait = Trait.toCommonFlags
```
and use it instead.

**Comment by smarter** [comment]
In `ModuleCompleter` we do:
```scala
denot.setFlag(from.flags.toTermFlags & RetainedModuleValFlags)
```
So I don't think that we can put `Lazy` in `FromStartFlags`.

**Comment by odersky** [comment]
Alternatively, we could have two fromStartFlags; one for terms and one for
types. That would be the cleaner solution, I think.

On Fri, Oct 23, 2015 at 11:03 PM, Guillaume Martres <
notifications@github.com> wrote:

> In ModuleCompleter we do:
>
> denot.setFlag(from.flags.toTermFlags & RetainedModuleValFlags)
>
> So I don't think that we can put Lazy in FromStartFlags.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/lampepfl/dotty/issues/850#issuecomment-150692098>.
>



-- 
Martin Odersky
EPFL

**Comment by smarter** [comment]
That's what I proposed above more or less but I wasn't sure if it was feasible:
> I see, so can we check if something is a type or a term without completing it? If so we could rename FromStartFlags to FromStartCommonFlags and add FromStartTypeFlags.

**Comment by DarkDimius** [comment]
```
denot.setFlag(from.flags.toTermFlags & RetainedModuleValFlags)
```
I'm not getting why this conflicts adding `Lazy` to `FromStartFlags`. Could you elaborate please?

**Comment by smarter** [comment]
Because it means we can set `Lazy` after the symbol was created, no?

**Comment by smarter** [comment]
Fixed by #860.
